worker/caasfirewaller/application_worker.go

```diff
@@ -0,0 +1,105 @@
+// Copyright 2017 Canonical Ltd.
+// Licensed under the AGPLv3, see LICENCE file for details.
+
+package caasfirewaller
+
+import (
+	"github.com/juju/errors"
+	"gopkg.in/juju/worker.v1"
+
+	"github.com/juju/juju/apiserver/params"
+	"github.com/juju/juju/caas"
+	"github.com/juju/juju/worker/catacomb"
+)
+
+type applicationWorker struct {
+	catacomb          catacomb.Catacomb
+	application       string
+	applicationGetter ApplicationGetter
+	serviceExposer    ServiceExposer
+
+	lifeGetter LifeGetter
+}
+
+func newApplicationWorker(
+	application string,
+	applicationGetter ApplicationGetter,
+	applicationExposer ServiceExposer,
+	lifeGetter LifeGetter,
+) (worker.Worker, error) {
+	w := &applicationWorker{
+		application:       application,
+		applicationGetter: applicationGetter,
+		serviceExposer:    applicationExposer,
+		lifeGetter:        lifeGetter,
+	}
+	if err := catacomb.Invoke(catacomb.Plan{
+		Site: &w.catacomb,
+		Work: w.loop,
+	}); err != nil {
+		return nil, errors.Trace(err)
+	}
+	return w, nil
+}
+
+// Kill is part of the worker.Worker interface.
+func (w *applicationWorker) Kill() {
+	w.catacomb.Kill(nil)
+}
+
+// Wait is part of the worker.Worker interface.
+func (w *applicationWorker) Wait() error {
+	return w.catacomb.Wait()
+}
+
+func (aw *applicationWorker) loop() error {
+	// TODO(caas) - get this from the application backend
+	config := caas.ResourceConfig{}
+	config[caas.JujuExternalHostNameKey] = "localhost"
+
+	appWatcher, err := aw.applicationGetter.WatchApplication(aw.application)
+	if err != nil {
+		if params.IsCodeNotFound(err) {
+			return nil
+		}
+		return errors.Trace(err)
+	}
+	if err := aw.catacomb.Add(appWatcher); err != nil {
+		return errors.Trace(err)
+	}
+
+	exposed, err := aw.applicationGetter.IsExposed(aw.application)
```

Comment:
So you're assuming that the initial state of the model reflects reality? That's not a sound assumption.

Rather than fetching the initial value here, I think a better approach would be:

```go
var previouslyExposed bool
initial := true
for {
    select {
    ...
    case _, ok := <-appWatcher.Changes():
        ...
        exposed, err := aw.applicationGetter.IsExposed(aw.application)
        ...
        if !initial && exposed == previouslyExposed {
            continue
        }
        ...
        initial = false
        previouslyExposed = exposed
    }
}
```

This way, you will always call ExposeService or UnexposeService once when the firewaller starts up. That's necessary in case of partial handling of a change. This means those methods need to deal with being called twice in a row. Alternatively, you'll need to introduce a third method to check whether or not the service is already exposed.

Reply:
Fair point. This was based on the mode of operation of the IAAS firewaller, so I think we need to consider fixing that one as well. Fixed here.

Reply:
The IAAS firewaller does the latter option: it calls an environ/instance method to get the currently-applied ingress rules, and then just applies the delta.